animations-dbus/animations-dbus-server-object.c

```diff
@@ -735,3 +744,69 @@ animations_dbus_server_new_with_connection_async (AnimationsDbusServerEffectFact
                               "effect-factory", factory,
                               NULL);
 }
+
+/**
+ * animations_dbus_server_stop:
+ * @self: an #AnimationsDbusServer
+ * @cancellable: (nullable): a #GCancellable, or %NULL
+ * @error: return location for a #GError, or %NULL
+ *
+ * Stop the server, unexport all D-Bus objects, and unown the well-known
+ * D-Bus name for the service.
+ *
+ * Once stopped, the server object can only be unreffed and destroyed: it cannot
+ * be restarted.
+ *
+ * Returns: %TRUE on success, %FALSE otherwise
+ */
+gboolean
+animations_dbus_server_stop (AnimationsDbusServer  *self,
+                             GCancellable          *cancellable,
+                             GError               **error)
+{
+  AnimationsDbusServerPrivate *priv = animations_dbus_server_get_instance_private (self);
+  gboolean error_seen = FALSE;
+
+  g_return_val_if_fail (ANIMATIONS_DBUS_IS_SERVER (self), FALSE);
+  g_return_val_if_fail (cancellable == NULL || G_IS_CANCELLABLE (cancellable), FALSE);
+  g_return_val_if_fail (error == NULL || *error == NULL, FALSE);
+
+  while (priv->animatable_surfaces != NULL && priv->animatable_surfaces->len > 0)
+    {
+      AnimationsDbusServerSurface *surface = g_ptr_array_index (priv->animatable_surfaces, 0);
+
+      if (!animations_dbus_server_unregister_surface (self, surface, NULL))
+        error_seen = TRUE;
+    }
+
+  if (priv->animation_managers != NULL)
+    {
+      /* Operate on a copy of the keys of @animation_managers to avoid iterating while modifying the hash table */
+      GList *names = g_hash_table_get_keys (priv->animation_managers);
```

Comment:
I dont think that is the case, but wonder if the contents of the keys can be modified once the hash is modified. From docs:
```
The content of the list is owned by the hash table and should not be modified or freed. Use g_list_free() when done using the list.
```

Alternatively we could use g_hash_table_get_keys_as_array which seems to always give us a copy?

Reply:
Uff, the return value of `g_hash_table_get_keys()` is of course `(transfer container)` and not `(transfer full)`. Fixed that.

The individual name strings should be OK to use even while the table is modified, since only the one which we’re calling `unsubscribe_client()` on should end up being freed, and that’s after this code is finished with it.

Reply:
`g_hash_table_get_keys_as_array()` can’t give us a copy as `GHashTable` doesn’t know how to deep-copy the key or value pointers it stores.

Reply:
Great, I had missed the `g_list_free_full()` issue, good catch.